impl/maven-impl/src/main/java/org/apache/maven/impl/PathSelector.java

```diff
@@ -0,0 +1,622 @@
+            if (pattern != null && !pattern.isEmpty()) {
+                if (pattern.indexOf(':') <= MAVEN_SYNTAX_THRESHOLD) {
+                    pattern = pattern.replace(File.separatorChar, '/');
+                    if (pattern.endsWith("/")) {
+                        pattern += "**";
+                    }
+                    // Following are okay only when "**" means "0 or more directories".
+                    while (pattern.endsWith("/**/**")) {
+                        pattern = pattern.substring(0, pattern.length() - 3);
+                    }
+                    while (pattern.startsWith("**/**/")) {
+                        pattern = pattern.substring(3);
+                    }
+                    pattern = pattern.replace("/**/**/", "/**/");
+                    pattern = pattern.replace("\\", "\\\\")
+                            .replace("[", "\\[")
+                            .replace("]", "\\]")
+                            .replace("{", "\\{")
+                            .replace("}", "\\}");
+                    normalized.add(DEFAULT_SYNTAX + pattern);
+                    /*
+                     * If the pattern starts or ends with "**", Java GLOB expects a directory level at
+                     * that location while Maven seems to consider that "**" can mean "no directory".
+                     * Add another pattern for reproducing this effect.
+                     */
+                    addPatternsWithOneDirRemoved(normalized, pattern, 0);
+                } else {
+                    normalized.add(pattern);
+                }
+            }
+        }
+        return simplify(normalized, excludes);
+    }
+
+    /**
+     * Adds all variants of the given pattern with {@code **} removed.
+     * This is used for simulating the Maven behavior where {@code "**} may match zero directory.
+     * Tests suggest that we need an explicit GLOB pattern with no {@code "**"} for matching an absence of directory.
+     *
+     * @param patterns where to add the derived patterns
+     * @param pattern  the pattern for which to add derived forms, without the "glob:" syntax prefix
+     * @param end      should be 0 (reserved for recursive invocations of this method)
+     */
+    private static void addPatternsWithOneDirRemoved(final Set<String> patterns, final String pattern, int end) {
+        final int length = pattern.length();
+        int start;
+        while ((start = pattern.indexOf("**", end)) >= 0) {
+            end = start + 2; // 2 is the length of "**".
+            if (end < length) {
+                if (pattern.charAt(end) != '/') {
+                    continue;
+                }
+                if (start == 0) {
+                    end++; // Ommit the leading slash if there is nothing before it.
+                }
+            }
+            if (start > 0) {
+                if (pattern.charAt(--start) != '/') {
+                    continue;
+                }
+            }
```

Comment:
```suggestion
            if (start > 0 && pattern.charAt(--start) != '/') {
                continue;
            }
```
could merge

Reply:
@desruisseaux ? you wan to include this suggestion ?

Reply:
I have no objection. I was waiting for other comments in order to merge them in a single commit.

Reply:
I committed, but actually the previous code was done that way for consistency with the previous `if` block. The two blocks were following the same pattern:

```
if (end < length) {
    // Do some check on the suffix.
}
if (start > 0) {
    // Do some check on the prefix.
}
```

The check on the prefix is simpler than the check on the suffix, which is why it could be rewritten. But we lost the symmetry with the suffix case.

Reply:
this if looks like if else but thats off-topic, 

thanks for compression.